index.js

```diff
@@ -39,91 +35,53 @@ PostgresDB.prototype.commit = function(collection, id, op, snapshot, options, ca
    * }
    * snapshot: PostgresSnapshot
    */
-  pg.connect(this.pg_config, function(err, client, done) {
-    if (err) {
-      done(client);
-      callback(err);
-      return;
-    }
-    function commit() {
-      client.query('COMMIT', function(err) {
-        done(err);
-        if (err) {
-          callback(err);
-        } else {
-          callback(null, true);
-        }
-      })
+    this.pool.connect((err, client, done) => {
+      if (err) {
+        done(client);
+        callback(err);
+        return;
+      }
+    /*const*/ var query = {
+      name: 'sdb-commit-op-and-snap',
+      text: `With snaps as (
+        Insert into snapshots (collection,doc_id,doc_type, version,data)
+        Select n.* From ( select $1 c, $2 d, $4 t, $3::integer v, $5::jsonb daa)
+        n 
+        where v = (select version+1 v from snapshots where collection = $1 and doc_id = $2 for update) or not exists (select 1 from snapshots where collection = $1 and doc_id = $2 for update)
+        On conflict(collection, doc_id) do update set version = $3, data = $5 , doc_type = $4
+        Returning version
+        ) 
+        Insert into ops (collection,doc_id, version,operation)
+        Select n.* From ( select $1 c, $2 t, $3::integer v, $6::jsonb daa)
+        n 
+        where (v = (select max(version)+1 v from ops where collection = $1 and doc_id = $2) or not exists (select 1 from ops where collection = $1 and doc_id = $2 for update)) and exists  (select 1 from snaps)
+        Returning version`,
```

Comment:
I found this query really hard to follow, even after reformatting and reindenting it. It's doing a lot of stuff, but I don't know if it's correct--could you explain (perhaps in a comment in the file) what this is doing and why?

Here's the reformatted version I made:
```sql
WITH snaps AS (
  INSERT INTO snapshots (collection, doc_id, doc_type, version, data)
  SELECT n.*
  FROM (SELECT $1 c, $2 d, $4 t, $3::integer v, $5::jsonb daa) n 
  WHERE v = (
    SELECT version+1 v
    FROM snapshots
    WHERE collection = $1 AND doc_id = $2
    FOR UPDATE
  ) OR NOT EXISTS (
    SELECT 1
    FROM snapshots
    WHERE collection = $1 AND doc_id = $2
    FOR UPDATE
  )
  ON CONFLICT (collection, doc_id) DO UPDATE SET version = $3, data = $5, doc_type = $4
  RETURNING version
)
INSERT INTO ops (collection, doc_id, version, operation)
SELECT n.* FROM (SELECT $1 c, $2 t, $3::integer v, $6::jsonb daa) n 
WHERE (
  v = (
    SELECT max(version)+1 v
    FROM ops
    WHERE collection = $1 AND doc_id = $2
  ) OR NOT EXISTS (
    SELECT 1
    FROM ops
    WHERE collection = $1 AND doc_id = $2
    FOR UPDATE
  )
) AND EXISTS (SELECT 1 FROM snaps)
RETURNING version
```

Reply:
Yeah it non-trivial sql. I've now added the following comment 

This query uses common table expression to upsert the snapshot table 
(iff the new version is exactly 1 more than the latest table or if
the document id does not exists)

It will then Insert into the ops table if it is exactly 1 more than the 
latest table or it the first operation and iff the previous insert into
the snapshot table is successful.

This result of this query the version of the newly inserted operation

If either the ops or the snapshot insert fails then 0 rows are returned

If 0 zeros are return then the callback must return false